### ISSUE TITLE
Allow JsonExtensionData on Records

### ISSUE THREAD

**Issue by danielcrabtree** [dotnet/runtime#69123]
Currently you cannot use JsonExtensionData on records. 

You can try as follows: ```record TestRecord([property:JsonExtensionData] Dictionary<string, JsonElement> Data);```

But you'll get this exception at runtime:
```
System.InvalidOperationException: The extension data property 'Data' on type 'TestRecord' cannot bind with a parameter in the deserialization constructor.
   at System.Text.Json.ThrowHelper.ThrowInvalidOperationException_ExtensionDataCannotBindToCtorParam(JsonPropertyInfo jsonPropertyInfo)
   ```

This exception is thrown in JsonTypeInfo.cs due to the record putting the JsonExtensionData property in the constructor. Is there any reason for this restriction on not having extension properties in the constructor? As this seems to severely limit their use on records.

It is of course possible to add a property to the record directly, but then that property is not part of the immutable contract of the record and makes it dangerous to use records in the way they are intended.

**Comment by eiriktsarpalis** [comment]
This is by design -- `JsonExtensionData` properties are special in that they accumulate all unbound JSON properties and as such they can only be populated once the entire underlying JSON object has been read. On the other hand, constructor-based deserialization will try to deserialize constructor parameters only and call the constructor as early as possible. This creates a dependency problem, which is why `JsonExtensionData` is explicitly not allowed as a constructor parameter.

It should still be possible to use `JsonExtensionData` by employing regular properties. This works even with positional records:
```C#
record TestRecord(int x)
{
    [JsonExtensionData]
    public Dictionary<string, JsonElement> Data { get; set; }
}
```

**Comment by danielcrabtree** [comment]
Yes, it works as a regular property, you can also use ```init``` instead of ```set``` to get immutable behavior.

Yes, I understand it is designed this way, but it seems to me as though it could be made to work with constructors as well, thereby allowing the positional record syntax. Obviously, there would be a performance overhead to using it, but that slower path would only apply when the constructor contained ```JsonExtensionData```. When that is not present, the fast path can be used. The performance in the slower case should be no different than when it is used as a property anyhow.

**Comment by eiriktsarpalis** [comment]
I don't think it's an issue of performance overhead, it's mostly a bootstrapping concern. You need to have the dictionary materialized _before_ the parent class gets constructed, which in most cases means that it needs to happen before the entire JSON object has been read. If we wanted to go the extra mile that might work with mutable dictionaries but it would make supporting read-only or immutable dictionaries impossible. 

Honestly though I don't think it's worth the churn and added complexity. There appears to be a workaround even for positional records.

**Comment by eiriktsarpalis** [comment]
Closing as answered. We don't believe adding support to `JsonExtensionData` for constructor parameters warrants the added complexity, given that there exist easy workarounds.
